indexer/ai/src/main/java/com/google/cloudsearch/ai/CustomSkill.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    /**
+     * Return the Inputs
+     * @return     Returns the Inputs for the skill in JSON format.
+     */
+    @Override
+    public JSONObject getInputs() {
+        return this.input;
+    }
+
+    /**
+     * Parse the JSON Object containing the inputs for the skill.
+     * @param input     Input in JSON format.
+     * @throws InvalidConfigException   Throws exception if the key in the input object is not expected.
+     */
+    @Override
+    public void parseInputs(JSONObject input) throws InvalidConfigException {
+        if(input == null) {
+            setInputs(new JSONObject());
+        }
+        else
+            setInputs(input);
+    }
+
+    /**
+     * No Filters are required for the Custom Skill.
+     * @param filter
+     */
+    @Override
+    public void setFilter(JSONObject filter) { return; }
+
+    /**
+     * Returns the filters.
+     * @return
+     */
+    @Override
+    public JSONObject getFilter() {
+        return null;
+    }
+
+    /**
+     * Filters not required for Custom AI Skill.
+     * @param filter
+     * @throws InvalidConfigException
+     */
+    @Override
+    public void parseFilter(JSONObject filter) throws InvalidConfigException {
+        if(filter != null) {
+            throw new InvalidConfigException("Filters not supported in Custom AI Skill.");
+        }
+    }
+
+    /**
+     * Set the Cloud Function URL.
+     * @param url   URL for invoking the Cloud Function.
+     */
+    public void setURL(String url) {
+        this.url = url;
+    }
+
+    /**
+     * Return the Cloud Function URL
+     * @return
+     */
+    public String getURL() {
+        return this.url;
+    }
+
+    /**
+     * Parses the Configuration for the Custom Skill.
+     * @param aiSkill        JSONObject for the skill
+     * @param schema         CloudSearch schema
+     * @throws InvalidConfigException
+     */
+    @Override
+    protected void parse(JSONObject aiSkill, JSONObject schema) throws InvalidConfigException {
+        parseAISkillName((String) aiSkill.get(Constants.CONFIG_SKILL_NAME));
+        parseOutputMappings((JSONArray) aiSkill.get(Constants.CONFIG_OUTPUT_MAPPINGS), schema);
+        parseInputs((JSONObject) aiSkill.get(Constants.CONFIG_INPUTS));
+        parseFilter((JSONObject) aiSkill.get(Constants.CONFIG_FILTERS));
+        setURL((String) aiSkill.get(Constants.CONFIG_CLOUD_FUNCTION_URL));
+    }
+
+    /**
+     *  No specific setup required.
+     */
+    @Override
+    public void setupSkill() {}
+
+    /**
+     * Execute the Cloud Function.
+     * JSON Response Format: { "key1" : [List of Values], "key2" : [List of Values] }
+     * key1 and key2 are output field names in the mapping.
+     * @param contentOrURI      The actual file content or Cloud storage URI
+     * @param structuredData    For storing CloudSearch Structured Data.
+     */
+    @Override
+    public void executeSkill(String contentOrURI, Multimap<String, Object> structuredData) {
+        try {
```

Comment:
why we have so many nested try catch blocks in this function ?

Reply:
Because for exceptions except IOException I wanted to throw invalid response exception, basically for everything related to the response.

Reply:
Isn't one try catch block enough for all of them

Reply:
Okay sure I will remove the nested try

Reply:
Check the latest commit in the PR